contracts/ModuleFactoryBase.sol

```diff
@@ -0,0 +1,73 @@
+//SPDX-License-Identifier: Unlicense
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts/utils/introspection/ERC165.sol";
+import "@openzeppelin/contracts/access/AccessControl.sol";
+import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import "./interfaces/IModuleFactory.sol";
+
+/// @notice An abstract contract to be inherited by module contracts
+abstract contract ModuleFactoryBase is
+    IModuleFactory,
+    ERC165,
+    AccessControl,
+    Initializable
+{
+    bytes32 public constant VERSION_ROLE = keccak256("VERSION_ROLE");
+    VersionInfo[] public versionControl;
+
+    /// @dev add a new version to update module users
+    /// @param _semanticVersion semantic version control
+    /// @param _frontendURI IPFS hash of the static frontend
+    /// @param _impl address of the impl
+    function addVersion(
+        string calldata _semanticVersion,
+        string calldata _frontendURI,
+        address _impl
+    ) external {
+        if (!hasRole(VERSION_ROLE, msg.sender)) {
+            revert NotAuthorized();
+        }
+        versionControl.push(VersionInfo(_semanticVersion, _frontendURI, _impl));
+        emit VersionCreated(_semanticVersion, _frontendURI, _impl);
+    }
+
+    /// @dev Creates a module
+    /// @param data The array of bytes used to create the module
+    /// @return address[] Array of the created module addresses
+    function create(bytes[] calldata data)
+        external
+        virtual
+        returns (address[] memory);
+
+    /// @dev Current version of system
+    /// @return currentVersion the current VersionInfo
+    function currentVersionInfo()
+        public
+        view
+        returns (VersionInfo memory currentVersion)
+    {
+        currentVersion = versionControl[versionControl.length - 1];
+    }
```

Comment:
Wondering if this needs to be something that's resolved on-chain, for similar reasons as previous comment. Why not just whatever client/frontend figure out what's relevant for them, based on the semantic versioning information stored in the `versionControl` array.